### ISSUE TITLE
redis cluster multi  bug

### ISSUE THREAD

**Issue by bournes** [phpredis/phpredis#876]
`$redis->multi();
$redis->hget('test1','a');   //cluster    nodes 1
$redis->hget('test2','a');   //cluster    nodes 1
print_r($redis->exec());
`
`PHP Fatal error:  Uncaught RedisClusterException: Error processing EXEC across the cluster in /root/testr.php:20
Stack trace:
#0 /root/testr.php(20): RedisCluster->exec()
#1 {main}
  thrown in /root/testr.php on line 20
`
------------------------------------------------------------------------------------------

`$redis->multi();
$redis->hget('test1','a');   //cluster    nodes 1
$redis->hget('test3','a');   //cluster    nodes 2
print_r($redis->exec());
`
is ok....

**Comment by tatekan** [comment]
Same issue #868

**Comment by michael-grunder** [comment]
This does seem to be a bug.  I'm at an offsite this week but will try to investigate during downtime

Cheers
Mike

**Comment by bournes** [comment]
@Mike  look at me.....

**Comment by bournes** [comment]
@michael-grunder  any  news about ?

**Comment by michael-grunder** [comment]
OK this is going to be a somewhat complicated bug to solve.  The problem is that within a MULTI ... EXEC block, all keys have to hash to the same *slot*.  It doesn't matter that they are actually on the same server.

It's going to take some designing to make it work, and is going to slow down MULTI...EXEC operations (quite a lot actually) but there isn't any way around it, as this is how Redis works.  :/

**Comment by tatekan** [comment]
Hi @michael-grunder,
I understand your point but somehow it works fine even if the keys are not in the same node.
I've ran several tests using MULTI...EXEC and I've always got back all the values even if the keys were on different nodes/slots.
I'm not sure if phpredis is "falling back" running single commands for those particular cases, but at least it doesn't break.

Thanks

**Comment by michael-grunder** [comment]
Hey,

The issue is that when I wrote the cluster `MULTI` code, I was running it against 3.0.x (and beta versions of Redis cluster).  At that point, you could group multiple keys that were housed on a given redis instance inside a `MULTI...EXEC` call.

This was actually an oversight, however and a fix was put in place to disallow keys which hashed to different individual slots from being requested inside the same `MULTI...EXEC` block.

Here is the original issue and bug fix in Redis proper:
https://github.com/antirez/redis/issues/2515

I need to update phpredis in such a way that it wraps MULTI ... EXEC calls _per slot_ instead of per node.

I'll update when I have more information for you

Cheers!
Mike

**Comment by Barbery** [comment]
+1, if have some way to solve it, please tell me, thank you

**Comment by michael-grunder** [comment]
@Barbery You can solve the issue in your php code by wrapping any keys with brackets, so they are forced to hash to the same slot.  You do that like this in Redis Cluster: `{this-is-hashed}-other-part`.

I know that's not a proper solution, but would allow you to force every key within a multi-exec block to hash to the same slot.

I am hoping to update the way phpredis handles this in the next few days though, and will update this thread when I have something working.

Cheers
Mike

**Comment by bournes** [comment]
Dear  @michael-grunder   .
       Have you solved the problem?  ....... Can I use the Redis 3.2.x ?

**Comment by bournes** [comment]
look at me　！

**Comment by jschneider98** [comment]
ping

**Comment by maxgalbu** [comment]
Is this solved? I am getting the same issue, using version 4.0.0 and php 5.6

**Comment by michael-grunder** [comment]
Hi everyone,

This problem still exists in phpredis but to be honest it's not clear to me what a "solution" would look like, or that this is a bug.

Redis now errors out if keys that map to different slots are accessed in the same MULTI block, which doesn't leave very many good options for phpredis.

There is no way to silently wrap MULTI ... EXEC calls in such a way to allow for the previous functionality without reordering operations and/or turning a single MULTI ... EXEC block into lots of MULTI ... EXEC blocks (which defeats the whole point of MULTI).

Consider a very simple example:

```bash
127.0.0.1:7000> set foo2 bar2
OK
127.0.0.1:7000> set foo6 bar6
OK
127.0.0.1:7000> multi
OK
127.0.0.1:7000> get foo2
QUEUED
127.0.0.1:7000> get foo6
QUEUED
127.0.0.1:7000> exec
(error) CROSSSLOT Keys in request don't hash to the same slot
```

Translated to phpredis:

```php
$obj_rc->multi()->get('foo2')->get('foo6')->exec();
```

The only potential solution I see is grouping MULTI ... EXEC calls _per slot_, but that would mean turning a single MULTI ... EXEC block into many (depending on how many slots we end up touching.

The problem with that solution is that it's no longer a single transaction but rather a bunch of sequential transactions and I think this will lead to confusion.

**Comment by maxgalbu** [comment]
I agree with what you said. What IMO would be quite useful is to allow pipelining commands, eg

```php
$redis = new Redis();
$pipe = $redis->multi(Redis::PIPELINE);
foreach ($users as $user_id => $username)
{
	$pipe->set('user:' . $user_id . ':username', $username);
}

// Executes all of the commands in one shot
$pipe->exec();
```

Pipeline still is quite useful and results in faster requests, when requesting multiple keys. We're currently testing corvus (https://github.com/eleme/corvus) on top of redis cluster, because it allows pipelining commands using the \Redis client (instead of \RedisCluster)

**Comment by michael-grunder** [comment]
Hey,

Not providing pipelining to the Redis cluster class was an intentional decision when I was first building it.  The issue as I saw it was that pipelining in a distributed system was dangerous given that an error could occur at any point in the process leaving the cluster in an inconsistent state.

That being said, I do see why it could be useful (e.g. if you're OK with risking an inconsistent state).  I could see adding pipelining to cluster at some point in the future, but there are a few more pressing issues I need to tackle first.

**Comment by maxgalbu** [comment]
By "inconsistent state" you mean "some keys could go missing", i guess?
Regarding reads instead, pipeline would really help when getting multiple keys, it's 4/5 times faster when compared to reading without transaction in redis cluster (0.20 ms for 750 keys with pipeline using corvus, 1 second without pipeline)

**Comment by michael-grunder** [comment]
Yeah, I mean that if we allow pipelining into cluster a failure in the middle of the process would occur on a node but the rest of the commands would continue to be executed.

Pipelining read-only commands to Redis cluster probably isn't as bad.

I'm glad to hear corvus is working for you though!  I will look into what it would take to enable pipelining for cluster.  I suppose it would be fine if the risks were made clear.

**Comment by JohannCarfantan** [comment]
Hey,

I'm facing the same problem, @michael-grunder do you have any update on this?
I understand that using brackets is a solution in a way but in my case this work is enormous and I'd like to avoid this. I'm also executing the same request that throws me this error in nodejs with the package ioredis (https://github.com/luin/ioredis) and there is no error, so it seems that they found a solution to this problem. I don't know if this can help, but I'd really like a fix on this.

Cheers

**Comment by michael-grunder** [comment]
I'm not sure there is really a fix for this.

There is no way to execute commands on keys in more than one slot within a _single_ MULTI call, because Redis will just plain refuse to do it.

This means that you lose transaction guarantees because PhpRedis would have to take your single `MULTI...EXEC` call and split it into multiple transactions, which is likely already broken for most use cases.

If order of operation doesn't matter, we could theoretically group keys by slot number (this would be more performant) but again, is probably broken from the outset.

Can you describe exactly what you're attempting to do.  I can think of some narrow use cases where, for example, it would be fine to reorder the commands but I doubt that's a high percentage of situations.

**Comment by JohannCarfantan** [comment]
I am facing two case when using `MULTI...EXEC` command : 

1. performing X times a same request in a for loop with a company id and a user id value changing in my keys. Here is a key example : `Enterprise:X:User:Y:conversations`
2. chaining some requests that have to be executed in an exact order. Example : `SADD` then `SMEMBERS`, or `HSET` then `HGETALL`, etc...

I think that I have to replace every call of the `MULTI...EXEC` by another way to realize this, which is a painful work in my case...

Here is how I will proceed

**First case** 

- before

```php
$redis->multi();
foreach($users as $user){
    $key = "Enterprise:" . $user->getEnterpriseId() .":User:" . $user->getId() . ":conversations";
    $redis->lRange($key, 0, 10);
}
$result = $redis->exec();
```

- after 

```php
$result = [];
foreach($users as $user){
    $key = "Enterprise:" . $user->getEnterpriseId() .":User:" . $user->getId() . ":conversations";
    array_push($result, $redis->lRange($key, 0, 10));
}
```

**Second case**

- before

```php
$redis->multi();
$redis->sadd("connectedUsers", $user->getId());
$redis->smembers("connectedUsers");
$result = $redis->exec();
```

- after

```php
$result = [];
array_push($result, $redis->sadd("connectedUsers", $user->getId()));
array_push($result, $redis->smembers("connectedUsers"));
```

**Conclusion**
I hope these modifications won't impact my code, and please if you see a problem tell me.
Actually I think that the main error I made was that I wanted to write my code the same way that I did in nodejs with ioredis client, and in most cases the usage of `MULTI...EXEC` was a mistake.

Cheers,

Johann

**Comment by TheLevti** [comment]
Encountering the same issue now.

Laravel test runs:

```php
$x = Cache::store()->getStore();

$y = $x->putMany([
    'a' => new Entity(),
    'b' => new Entity(),
    'c' => new Entity(),
], 60);

dump($x->get('b'));

dump($y);
```

Redis monitor returns:

```
1651067707.409569 [0 10.150.2.3:53400] "MULTI"
1651067707.409947 [0 10.150.2.3:53400] "DISCARD"
```

Phpunit returns:

```
There was 1 error:

1) Tests\Http\MiddleWare\XXXXX\XXXXMiddlewareTest::it_tests_something
RedisClusterException: Error processing EXEC across the cluster

/var/www/html/vendor/illuminate/redis/Connections/Connection.php:116
/var/www/html/vendor/illuminate/redis/Connections/PhpRedisConnection.php:533
/var/www/html/vendor/illuminate/redis/Connections/Connection.php:220
/var/www/html/vendor/illuminate/redis/Connections/PhpRedisConnection.php:575
/var/www/html/vendor/illuminate/cache/RedisStore.php:124
/var/www/html/tests/Http/MiddleWare/XXXXX/XXXXMiddlewareTest.php:38

ERRORS!
Tests: 1, Assertions: 0, Errors: 1.
```

**Comment by TheLevti** [comment]
We see different kind of errors depending also on whether we have slot caching enabled.

**Comment by clarence-eldefors** [comment]
I just ran into this myself and found a workaround that seems to be working. It is to make sure that all keys requested gets the same slot by changing what it hashes. Redis cluster documents it's hashing at https://redis.io/docs/reference/cluster-spec/#hash-tags

Basically by wrapping some identifier in {} you will get those keys mapped to the same slot. An example from above would be to change `Enterprise:` to `{Enterprise}` to be able to use `multi()` on all Enterprise entries.

Whereas this would be rather easy to solve in many frameworks or ORMs I am not sure there's a straight forward way to change this in phpredis. But the phpredis documentation does suggest it has better support for `multi()` for redis cluster than what it actually has. So at least a documentation update would be good?
